day7/part2.go

```diff
@@ -0,0 +1,196 @@
+package day7
+
+import (
+	"fmt"
+	"ryepup/advent2021/utils"
+	"sync"
+)
+
+/*
+The crabs don't seem interested in your proposed solution. Perhaps you
+misunderstand crab engineering?
+
+As it turns out, crab submarine engines don't burn fuel at a constant rate.
+Instead, each change of 1 step in horizontal position costs 1 more unit of fuel
+than the last: the first step costs 1, the second step costs 2, the third step
+costs 3, and so on.
+
+As each crab moves, moving further becomes more expensive. This changes the best
+horizontal position to align them all on; in the example above, this becomes 5:
+
+    Move from 16 to 5: 66 fuel
+    Move from 1 to 5: 10 fuel
+    Move from 2 to 5: 6 fuel
+    Move from 0 to 5: 15 fuel
+    Move from 4 to 5: 1 fuel
+    Move from 2 to 5: 6 fuel
+    Move from 7 to 5: 3 fuel
+    Move from 1 to 5: 10 fuel
+    Move from 2 to 5: 6 fuel
+    Move from 14 to 5: 45 fuel
+
+This costs a total of 168 fuel. This is the new cheapest possible outcome; the
+old alignment position (2) now costs 206 fuel instead.
+
+Determine the horizontal position that the crabs can align to using the least
+fuel possible so they can make you an escape route! How much fuel must they
+spend to align to that position?
+*/
+func Part2(path string) (int, error) {
+	return Part2Opts(path, Naive, ForLoop) // fastest approach on my machine
+}
+
+func Part2Opts(path string, cache CacheStrategy, proc ParallelStrategy) (int, error) {
+	if cache == Naive && proc != ForLoop {
+		return 0, fmt.Errorf("invalid strategy combination")
+	}
+
+	positions, err := utils.ReadIntCsv(path)
+	if err != nil {
+		return 0, err
+	}
+	maxPosition := utils.MaxInt(positions...)
+	solutions := make([]int, maxPosition+1)
+	costFunction := makeFuelStrategy(cache)
+	processor := makeProcessor(proc)
+
+	processor(positions, maxPosition, costFunction, solutions)
+
+	return utils.MinInt(solutions...), nil
+}
+
+func rawFuelCost(distance int) int {
+	/*
+		1 -> 1: 1
+		2 -> 3: 1 + 2
+		3 -> 6: 1 + 2 + 3
+		4 -> 10: 1 + 2 + 3 + 4
+
+		TODO: there's probably a cheaper / clever math trick to do this
+	*/
+	cost := 0
+
+	for i := 1; i <= distance; i++ {
+		cost += i
+	}
+	return cost
```

Comment:
`return distance * (distance + 1) / 2`

The intuition I learned is that if you are summing from, for example 1 to 10, create a set of pairs that are equal to `n+1`: `10 + 1 = 11`, `9 + 2 = 11`, `8 + 3 = 11`.  How many of these pairs do you have? `n/2` of them, or 5 in this example.

https://en.wikipedia.org/wiki/1_%2B_2_%2B_3_%2B_4_%2B_%E2%8B%AF

Reply:
nice, that was bugging me. Makes a nice change to the benchmarks, too.